draco/synthesis/mockcatalog.py

```diff
@@ -0,0 +1,1045 @@
+            # is the voxel size along the z direction.
+            from cora.util import cosmology
+
+            cosmo = cosmology.Cosmology()
+            z_weights_global = np.zeros(gs, dtype=np.float64)
+
+            # First, we compute the normalization for each channel
+            # globally
+            for fi, freq in enumerate(source_map.index_map["freq"]):
+                z_min = units.nu21 / (freq[0] + 0.5 * freq[1]) - 1
+                z_max = units.nu21 / (freq[0] - 0.5 * freq[1]) - 1
+                z_mean = units.nu21 / freq[0] - 1
+
+                z_weights_global[fi] = cosmo.comoving_distance(z_mean) ** 2 * (
+                    cosmo.comoving_distance(z_max) - cosmo.comoving_distance(z_min)
+                )
+
+            z_weights_global /= z_weights_global.sum()
+
+            # Select local section of weights
+            z_weights = mpiarray.MPIArray.wrap(z_weights_global[lo : lo + ls], axis=0)
+
+        # Create PDF map
+        pdf_map = self.make_pdf_map(source_map, z_weights)
+
+        return pdf_map
+
+
+class MockCatGenerator(task.SingleTask):
+    """Take PDF maps generated by task :class:`PdfGenerator`
+    and use it to draw mock catalogs.
+
+    Attributes
+    ----------
+    nsources : int
+        Number of sources to draw in each mock catalog.
+    ncats : int
+        Number of catalogs to generate.
+    sigma_z : float, optional
+        Standard deviation of Gaussian redshift errors (default: None)
+    sigma_z_over_1plusz : float, optional
+        Standard deviation of Gaussian redshift errors will be set to
+        this parameter times (1+z). Only one of this and `sigma_z` can
+        be specified. Default: None
+    z_at_channel_centers : bool, optional
+        Place each source at a redshift corresponding to the center of
+        its frequency channel (True), or randomly distribute each source's
+        redshift within its channel (False). Default: False.
+    srcs_at_pixel_centers : bool, optional
+        Place each source precisely at Healpix pixel center (True), or
+        randomly distribute each source within pixel (False).
+        Default: False.
+    """
+
+    nsources = config.Property(proptype=int)
+    ncats = config.Property(proptype=int)
+
+    sigma_z = config.Property(proptype=float, default=None)
+    sigma_z_over_1plusz = config.Property(proptype=float, default=None)
+
+    z_at_channel_centers = config.Property(proptype=bool, default=False)
+    srcs_at_pixel_centers = config.Property(proptype=bool, default=False)
+
+    def setup(self, pdf_map):
+        """Pre-load information from PDF.
+
+        Parameters
+        ----------
+        pdf_map : :class:`containers.Map`
+            PDF from which to draw positions of sources.
+        """
+
+        # Get PDF map container and corresponding healpix Nside
+        self.pdf = pdf_map
+        self.nside = self.pdf.nside
+
+        # Get MPI communicator and rank
+        self.comm_ = self.pdf.comm
```

Comment:
Just use `self.comm` everywhere. It should be the same. I think the communicator the task gets is checked for consistency with what the data has.